opal/mca/shmem/shmem_types.h

```diff
@@ -103,7 +104,7 @@ struct opal_shmem_seg_hdr_t {
     opal_atomic_lock_t lock;
     /* pid of the segment creator */
     pid_t cpid;
-};
+} __opal_attribute_aligned__(OPAL_ALIGN_MIN);
```

Comment:
I'm not sure I understand what you mean here. The compiler has to pad the structure in order to guarantee proper alignment of object in an array (provided the array itself is properly aligned):

> Note that [...] the compiler’s ability to select a time-efficient alignment is primarily useful only when you plan to create arrays of variables having the relevant (efficiently aligned) type.

 A simple test to illustrate that:

```C
#include <stdio.h>

struct A {
  short a;
  char  b;
};

struct B {
  short a;
  char  b;
} __attribute__((aligned(16)));

int main()
{
  printf("%d vs %d\n", sizeof(struct A), sizeof(struct B));
  return 0;
}
```
GCC, ICC, and Clang yield:
```
$ ./a.out 
4 vs 16
```

In our case, the memory itself is page-aligned and the alignment attribute makes sure that the structure is properly padded to align the pointer returned to the user.

Reply:
You're right, the incurring padding changes the sizeof the struct changes to the specified alignment. Yesterday, I made a test case similar to your but instead of aligning the struct I aligned the typedef, and it showed a differnet result.